consensus/src/core.rs

```diff
@@ -174,52 +186,74 @@ impl<Mempool: 'static + NodeMempool> Core<Mempool> {
             return Ok(());
         }
 
+        // Ensure the vote is well formed.
+        vote.verify(&self.committee)?;
+
         // Add the new vote to our aggregator and see if we have a quorum.
-        if let Some(quorum) = self.aggregator.add_vote(vote.clone())? {
-            let qc = QC {
-                hash: vote.hash.clone(),
-                round: vote.round,
-                votes: quorum,
-            };
+        if let Some(qc) = self.aggregator.add_vote(vote.clone())? {
             debug!("Assembled {:?}", qc);
-            if !qc.timeout() {
-                self.update_high_qc(&qc);
+
+            // Process the QC.
+            self.process_qc(&qc).await;
+
+            // Make a new block if we are the next leader.
+            if self.name == self.leader_elector.get_leader(self.round) {
+                self.generate_proposal(None).await?;
             }
+        }
+        Ok(())
+    }
+
+    async fn handle_timeout(&mut self, timeout: &Timeout) -> ConsensusResult<()> {
+        debug!("Processing {:?}", timeout);
+        if timeout.round < self.round {
+            return Ok(());
+        }
+
+        // Ensure the timeout is well formed.
+        timeout.verify(&self.committee)?;
+
+        // Process the QC embedded in the timeout.
+        self.process_qc(&timeout.high_qc).await;
+
+        // Add the new vote to our aggregator and see if we have a quorum.
+        if let Some(tc) = self.aggregator.add_timeout(timeout.clone())? {
```

Comment:
If you get a TC you should broadcast it to ensure liveness

Reply:
Note: This is also necessary in 3-chain HS!